go/consensus/tendermint/abci/prune.go

```diff
@@ -60,111 +59,99 @@ type PruneConfig struct {
-	Prune(latestVersion int64)
-
-	// Subscribe subscribes to prune events.
-	Subscribe() (<-chan int64, *pubsub.Subscription, error)
+	Prune(ctx context.Context, latestVersion uint64) error
 }
 
 type statePrunerInitializer interface {
-	Initialize(latestVersion int64) error
+	Initialize(latestVersion uint64) error
 }
 
 type nonePruner struct{}
 
-func (p *nonePruner) Prune(latestVersion int64) {
+func (p *nonePruner) Prune(ctx context.Context, latestVersion uint64) error {
 	// Nothing to prune.
-}
-
-func (p *nonePruner) Subscribe() (<-chan int64, *pubsub.Subscription, error) {
-	return nil, nil, nil
+	return nil
 }
 
 type genericPruner struct {
 	logger *logging.Logger
-	tree   *iavl.MutableTree
-
-	eldestRetained int64
-	keepN          int64
+	ndb    nodedb.NodeDB
 
-	notifier *pubsub.Broker
+	earliestVersion uint64
+	keepN           uint64
 }
 
-func (p *genericPruner) Initialize(latestVersion int64) error {
+func (p *genericPruner) Initialize(latestVersion uint64) error {
 	// Figure out the eldest version currently present in the tree.
-	if p.eldestRetained = p.tree.EldestVersion(); p.eldestRetained == -1 {
-		// The tree is empty, nothing to purge.
-		p.eldestRetained = 0
-		return nil
+	var err error
+	if p.earliestVersion, err = p.ndb.GetEarliestRound(context.Background()); err != nil {
+		return fmt.Errorf("failed to get earliest round: %w", err)
 	}
 
-	return p.doPrune(latestVersion)
+	return p.doPrune(context.Background(), latestVersion)
 }
 
-func (p *genericPruner) Prune(latestVersion int64) {
-	if err := p.doPrune(latestVersion); err != nil {
+func (p *genericPruner) Prune(ctx context.Context, latestVersion uint64) error {
+	if err := p.doPrune(ctx, latestVersion); err != nil {
 		p.logger.Error("Prune",
 			"err", err,
 		)
-		panic(err)
+		return err
 	}
+	return nil
 }
 
-func (p *genericPruner) doPrune(latestVersion int64) error {
+func (p *genericPruner) doPrune(ctx context.Context, latestVersion uint64) error {
 	if latestVersion < p.keepN {
 		return nil
 	}
 
 	p.logger.Debug("Prune: Start",
 		"latest_version", latestVersion,
-		"start_version", p.eldestRetained,
+		"start_version", p.earliestVersion,
 	)
 
 	preserveFrom := latestVersion - p.keepN
-	for i := p.eldestRetained; i <= latestVersion; i++ {
-		if p.tree.VersionExists(i) {
-			if i >= preserveFrom {
-				p.eldestRetained = i
-				break
-			}
-
-			p.logger.Debug("Prune: Delete",
-				"latest_version", latestVersion,
-				"pruned_version", i,
-			)
+	for i := p.earliestVersion; i <= latestVersion; i++ {
+		if i >= preserveFrom {
+			p.earliestVersion = i
+			break
+		}
 
-			if err := p.tree.DeleteVersion(i); err != nil {
-				return err
-			}
+		p.logger.Debug("Prune: Delete",
+			"latest_version", latestVersion,
+			"pruned_version", i,
+		)
 
-			p.notifier.Broadcast(i)
+		err := p.ndb.Prune(ctx, i)
+		switch err {
+		case nil:
+		case nodedb.ErrNotEarliest:
+			p.logger.Debug("Prune: skipping non-earliest version",
+				"version", i,
+			)
+			continue
```

Comment:
In case the given version is earlier than the earliest.

I think the only time this happens is when the pruner is initialized with an empty version which is zero. But then a new higher version is created (as the first version can be arbitrary), so this would keep on failing.

Actually now that I've explained it it seems that the pruner should just detect this empty case and re-check what the earliest version is.